service/history/timerGate.go

```diff
@@ -125,20 +130,28 @@ func (timerGate *LocalTimerGateImpl) FireAfter(now time.Time) bool {
 // Update update the timer gate, return true if update is a success
 // success means timer is idle or timer is set with a sooner time to fire
 func (timerGate *LocalTimerGateImpl) Update(nextTime time.Time) bool {
+	// NOTE: negative duration will make the timer fire immediately
 	now := time.Now()
 
-	if !timerGate.nextWakeupTime.After(now) || timerGate.nextWakeupTime.After(nextTime) {
-		// if timer will not fire or next wake time is after the "next"
-		// then we need to update the timer to fire
+	if timerGate.timer.Stop() {
```

Comment:
Can we write a bench test to  measure TimerGate.Update with and without this change.

Reply:
done, here is the result:

after the change:
```
cadence|timer-bug ⇒ go test -cpu=1 -benchtime=10s -benchmem -run=^$ github.com/uber/cadence/service/history -bench ^BenchmarkLocalTimer$
goos: darwin
goarch: amd64
pkg: github.com/uber/cadence/service/history
BenchmarkLocalTimer 	100000000	       107 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/uber/cadence/service/history	10.931s
cadence|timer-bug ⇒ go test -cpu=1 -benchtime=10s -benchmem -run=^$ github.com/uber/cadence/service/history -bench ^BenchmarkLocalTimer$
goos: darwin
goarch: amd64
pkg: github.com/uber/cadence/service/history
BenchmarkLocalTimer 	100000000	       100 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/uber/cadence/service/history	10.205s
cadence|timer-bug ⇒ go test -cpu=1 -benchtime=10s -benchmem -run=^$ github.com/uber/cadence/service/history -bench ^BenchmarkLocalTimer$
goos: darwin
goarch: amd64
pkg: github.com/uber/cadence/service/history
BenchmarkLocalTimer 	100000000	       108 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/uber/cadence/service/history	10.945s
cadence|timer-bug ⇒ go test -cpu=4 -benchtime=10s -benchmem -run=^$ github.com/uber/cadence/service/history -bench ^BenchmarkLocalTimer$
goos: darwin
goarch: amd64
pkg: github.com/uber/cadence/service/history
BenchmarkLocalTimer-4   	100000000	       211 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/uber/cadence/service/history	21.347s
cadence|timer-bug ⇒ go test -cpu=4 -benchtime=10s -benchmem -run=^$ github.com/uber/cadence/service/history -bench ^BenchmarkLocalTimer$
goos: darwin
goarch: amd64
pkg: github.com/uber/cadence/service/history
BenchmarkLocalTimer-4   	100000000	       193 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/uber/cadence/service/history	19.484s
cadence|timer-bug ⇒ go test -cpu=4 -benchtime=10s -benchmem -run=^$ github.com/uber/cadence/service/history -bench ^BenchmarkLocalTimer$
goos: darwin
goarch: amd64
pkg: github.com/uber/cadence/service/history
BenchmarkLocalTimer-4   	100000000	       236 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/uber/cadence/service/history	23.771s
```

before the change:
```
cadence|master ⇒ go test -cpu=1 -benchtime=10s -benchmem -run=^$ github.com/uber/cadence/service/history -bench ^BenchmarkLocalTimer$
goos: darwin
goarch: amd64
pkg: github.com/uber/cadence/service/history
BenchmarkLocalTimer 	200000000	        88.8 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/uber/cadence/service/history	26.831s
cadence|master⚡ ⇒ go test -cpu=1 -benchtime=10s -benchmem -run=^$ github.com/uber/cadence/service/history -bench ^BenchmarkLocalTimer$
goos: darwin
goarch: amd64
pkg: github.com/uber/cadence/service/history
BenchmarkLocalTimer 	200000000	        86.9 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/uber/cadence/service/history	26.456s
cadence|master⚡ ⇒ go test -cpu=1 -benchtime=10s -benchmem -run=^$ github.com/uber/cadence/service/history -bench ^BenchmarkLocalTimer$
goos: darwin
goarch: amd64
pkg: github.com/uber/cadence/service/history
BenchmarkLocalTimer 	200000000	        88.1 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/uber/cadence/service/history	26.589s
cadence|master⚡ ⇒ go test -cpu=4 -benchtime=10s -benchmem -run=^$ github.com/uber/cadence/service/history -bench ^BenchmarkLocalTimer$
goos: darwin
goarch: amd64
pkg: github.com/uber/cadence/service/history
BenchmarkLocalTimer-4   	100000000	       183 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/uber/cadence/service/history	18.578s
cadence|master⚡ ⇒ go test -cpu=4 -benchtime=10s -benchmem -run=^$ github.com/uber/cadence/service/history -bench ^BenchmarkLocalTimer$
goos: darwin
goarch: amd64
pkg: github.com/uber/cadence/service/history
BenchmarkLocalTimer-4   	100000000	       185 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/uber/cadence/service/history	18.653s
cadence|master⚡ ⇒ go test -cpu=4 -benchtime=10s -benchmem -run=^$ github.com/uber/cadence/service/history -bench ^BenchmarkLocalTimer$
goos: darwin
goarch: amd64
pkg: github.com/uber/cadence/service/history
BenchmarkLocalTimer-4   	100000000	       179 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/uber/cadence/service/history	18.042s
```

the performance impact is ~<20% latency increase, which is negligible